llvm/lib/Target/X86/X86ISelLowering.cpp

```diff
@@ -41273,6 +41273,33 @@ static SDValue combineTargetShuffle(SDValue N, const SDLoc &DL,
 
     return SDValue();
   }
+  case X86ISD::VPERMV3: {
+    // VPERM[I,T]2[B,W] are 3 uops on Skylake and Icelake so we try to use
+    // VPERMV.
+    if (VT.is512BitVector() || (VT.is256BitVector() && !Subtarget.hasEVEX512()))
+      return SDValue();
+    MVT ElementVT = VT.getVectorElementType();
+    if (ElementVT != MVT::i8 && ElementVT != MVT::i16)
+      return SDValue();
```

Comment:
Although VPERM2B/W are slower, do would not still avoid an extract subvector by widening D/Q/PS/PD cases?

Reply:
But we are also trying to avoid generating 512-bit instructions. I think I can put it in a separate patch in case it has negative effects.

Reply:
Can we try to scale the `i8`/`i16` cases?

Reply:
Do you mean to turn them into `i32`? I think it relies on specific mask patterns, I'm not sure if we have done that or not. Maybe worth a try if not.

Reply:
It does rely on specific mask patterns (see `scaleShuffleElements`).

Reply:
> Although VPERM2B/W are slower, do would not still avoid an extract subvector by widening D/Q/PS/PD cases?

#97206